### PR TITLE
My hw 02

### PR DESCRIPTION
**Pull request by NeverovDS**
My 3 tasks.For search my errors

### CODE REVIEW

File: HW02/Dzmitry_Neverov/task3/time_calculator.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      difference = (arr[1] - arr[0]).to_s
+      puts(difference)
+    else
+      #   arr_for_difference = []
+      arr[1..arr.size].map!.with_index(1) do |time, i|
+        (time - arr[i - 1]).abs
+      end
+      puts("arr_for_difference #{arr}")
+    end
+  end
```
Comment (vladhilko) [inline HW02/Dzmitry_Neverov/task3/time_calculator.rb]: Can you try this one?

```suggestion
  def calc_differ(arr)
      arr[1..arr.size].map!.with_index(1) do |time, i|
        (time - arr[i - 1]).abs
      end
  end
  
  puts calc_differ(arr)
```

File: HW01/Dzmitry_Neverov/dzmitry_neverov_hw_01_t_01.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+def calc_angle(hours, minutes)
+  angel = (minutes * 6 - (hours * 60 + minutes) * 0.5).abs
+  if angel > 180
+    360 - angel.abs
+  else
+    angel.abs
+  end
+end
+
+puts "#{calc_angle(hours, minutes)} degrees"
```
Comment (vladhilko) [inline HW01/Dzmitry_Neverov/dzmitry_neverov_hw_01_t_01.rb]: Can you remove HW1 from this PR? 😅

File: HW02/Dzmitry_Neverov/task1/handler.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+logs = File.readlines("#{__dir__}/logs_hw_02_t_01.txt")
+
+def task_1(logs)
+  logs.select do |line|
+    line.downcase =~ /.*error.*/
```
Comment (vladhilko) [inline HW02/Dzmitry_Neverov/task1/handler.rb]: Should return only the first line with an error

File: HW02/Dzmitry_Neverov/task2/logs_reader.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class ReaderLogs
```
Comment (vladhilko) [inline HW02/Dzmitry_Neverov/task2/logs_reader.rb]:
```suggestion
class LogsReader
```

File: HW02/Dzmitry_Neverov/task2/logs_reader.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  attr_reader :filename
+
+  def initialize(filename)
+    @filename = filename
+  end
+
+  def read_file
+    File.open(filename).readlines
```
Comment (vladhilko) [inline HW02/Dzmitry_Neverov/task2/logs_reader.rb]: 👍

File: HW02/Dzmitry_Neverov/task2/logs_validator.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  attr_reader :logs_to_check
+
+  def initialize(logs_to_check)
+    @logs_to_check = logs_to_check
+  end
+
+  LOG_VALIDATE = %r{
+(?<ip>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}).{5}\[
+(?<date>\d{1,2}/\w{3}/\d{4}\:\d{,2}:\d{,2}:\d{,2}\s\+\d{4})\]\s"POST\s
+(?<url>/\w+/\d+/\w+)\s
```
Comment (vladhilko) [inline HW02/Dzmitry_Neverov/task2/logs_validator.rb]: 👍

File: HW02/Dzmitry_Neverov/task2/start.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+require_relative 'logs_reader.rb'
+require_relative 'logs_validator.rb'
+
+logs_to_check = ReaderLogs.new(ARGV[0]).read_file
```
Comment (vladhilko) [inline HW02/Dzmitry_Neverov/task2/start.rb]:
```suggestion
logs = ReaderLogs.new(ARGV[0]).read_file
```

File: HW02/Dzmitry_Neverov/task2/logs_validator.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+  def task_2
+    valid_logs = 0
+    logs_to_check.map do |line|
+      next unless line.match?(LOG_VALIDATE)
+
+      matches = line.match(LOG_VALIDATE)
+
+      puts "#{matches[:date]} FROM: #{matches[:ip]} TO: #{matches[:url].upcase}"
+      valid_logs += 1
+    end
```
Comment (vladhilko) [inline HW02/Dzmitry_Neverov/task2/logs_validator.rb]: This method just should return arrays of new strings, `puts` should be outside

File: HW02/Dzmitry_Neverov/task2/logs_validator.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  LOG_VALIDATE = %r{
+(?<ip>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}).{5}\[
+(?<date>\d{1,2}/\w{3}/\d{4}\:\d{,2}:\d{,2}:\d{,2}\s\+\d{4})\]\s"POST\s
+(?<url>/\w+/\d+/\w+)\s
+}x.freeze
+
+  def task_2
+    valid_logs = 0
+    logs_to_check.map do |line|
+      next unless line.match?(LOG_VALIDATE)
```
Comment (vladhilko) [inline HW02/Dzmitry_Neverov/task2/logs_validator.rb]: Try to implement the following interface 
```suggestion
  def task_2
    valid_logs.map { .. }
  end
  
   private 
   def valid_logs
     logs_to_check.select { ... }
   end
```

File: HW02/Dzmitry_Neverov/task3/line_search.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class LineSearcher
+  attr_reader :some_lines
+
+  def initialize(some_lines)
+    @some_lines = some_lines
+  end
+
+  NEED_LOG = /.*Calling core with action.*/.freeze
```
Comment (vladhilko) [inline HW02/Dzmitry_Neverov/task3/line_search.rb]: constants should be directly under the class name 
```suggestion
class LineSearcher
  NEED_LOG = /.*Calling core with action.*/.freeze
```

File: HW02/Dzmitry_Neverov/task3/reader_logs.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
+class ReaderLogs
```
Comment (B0RlS) [inline HW02/Dzmitry_Neverov/task3/reader_logs.rb]: It is better to call the class LogsReader and not ReaderLogs

File: HW02/Dzmitry_Neverov/task3/time_calculator.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    if arr.length > 2
+      check_length(arr)
+    else
+      puts((arr[1] - arr[0]).to_s)
+    end
+  end
+
+  def time_one?(arr_for_times)
+    arr_for_times.length == 1
+  end
```
Comment (B0RlS) [inline HW02/Dzmitry_Neverov/task3/time_calculator.rb]: Better to make these classes private.

File: HW02/Dzmitry_Neverov/task3/time_calculator.rb [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+  def task_3
+    arr_for_times = []
+    search_time.each { |time| arr_for_times << Time.parse(time) }
+    if time_one?(arr_for_times).trust
```
Comment (B0RlS) [inline HW02/Dzmitry_Neverov/task3/time_calculator.rb]: Why are you making a method to check the length? You can insert a check in the condition if
```suggestion
    if arr_for_times.length == 1
    ....
```

File: HW02/Dzmitry_Neverov/task3/time_calculator.rb [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+  def search_time
+    found_rows.each { |rows| rows.match(TIME_FORMAT) }
+  end
+
+  def check_length(arr)
+    arr_for_difference = []
+    arr[1..arr.size].map!.with_index(1) do |time, i|
+      arr_for_difference << (time - arr[i - 1]).abs
+    end
+    puts("arr_for_difference #{arr_for_difference}")
```
Comment (B0RlS) [inline HW02/Dzmitry_Neverov/task3/time_calculator.rb]: It is bad practice to combine logic and information output.

**Pull request merged**